src/realmd/AuthSocket.cpp

```diff
@@ -473,13 +474,20 @@ bool AuthSocket::_HandleLogonChallenge()
                     if (!_token.empty() && _build >= 8606) // authenticator was added in 2.4.3
                         securityFlags = SECURITY_FLAG_AUTHENTICATOR;
 
+                    if (!_token.empty() && _build <= 6141)
+                        securityFlags = SECURITY_FLAG_PIN;
+
                     pkt << uint8(securityFlags);                    // security flags (0x0...0x04)
 
                     if (securityFlags & SECURITY_FLAG_PIN)          // PIN input
                     {
-                        pkt << uint32(0);
-                        pkt << uint64(0);
-                        pkt << uint64(0);
+                        uint32 gridSeedPkt = m_gridSeed = static_cast<uint32>(0/*urand()*/);
```

Comment:
Leaving this as 0 instead of urand() (neither of which really need a static_cast, but the compiler will optimize that out anyway) means that the keypad on the client is not scrambled. Adding any other number (or urand()) here will scramble the keypad, making text entry more difficult for no real benefit. It's not a physical keypad where an attacker could look at your finger movements to determine the code you're entering, after all. If an attacker can see your screen, you have bigger problems.

Left it in in case someone sees the need to enable scrambling, tho.

Reply:
It would be interesting to check which value makes the keypad have the buttons in order from top-left to bottom-right (the way old phones used to have them).